### ISSUE TITLE
Difference in saturation behavior between single-channel and multi-channel matrix operations

### ISSUE THREAD

**Issue by jkanem** [opencv/opencv#26138]
### System Information

OpenCV version 4.9.0
Operating System / Platform: Windows 10
Compiler & compiler version: MSVC2019

### Detailed description

For a single channel matrix `m`, an integer `c` and a cv::Scalar `b`, the operation
```
m = c*(m - b)
```
results in saturation being applied only after the right-hand-side has been fully evaluated. But for a multi-channel image, saturation is applied to `c*m` before subtracting `c*b`.

This leads to unexpected results. Let `c` be 4, let `m` be an 8-Bit 3-channel matrix (CV_8UC3) and have a value of 100 for all pixels and channels, and let `b` have a value of 70 for each channel. The result of `c*(m-b)` should be a matrix with values of `4*(100 - 70) = 120`. Instead it is a matrix of zeros because `4*100` was capped at 255 and after subtracting `4*70` we get -25 which becomes 0 after the final saturation operation. 

The accepted answer to this issue on SO (https://stackoverflow.com/a/78963620/5398144) traces this to `MatOp_AddEx::assign` in core/src/matrix_expressions.cpp (https://github.com/opencv/opencv/blob/4.x/modules/core/src/matrix_expressions.cpp#L1342). 
A 3-channel matrix operation, with the addition of a 3-channel scalar fall into the final `else` statement:
```
else
    {
        e.a.convertTo(dst, e.a.type(), e.alpha);
        cv::add(dst, e.s, dst);
    }
```
So the operations that are happening boil down to:
```
m.convertTo(dst, m.type(), c); // which applies saturation to c*m
cv::add(dst, -c*b, dst);
```

### Steps to reproduce

```
#include <iostream>
#include <opencv2/core/core.hpp>

int main()
{
  std::cout << "c m1 m3" << std::endl;
  for (int c = 1; c <= 9; c++){
    cv::Mat m1 = cv::Mat::zeros(1,1, CV_8UC1);
    cv::Mat m3 = cv::Mat::zeros(1,1, CV_8UC3);
    m1 += cv::Scalar(100);
    m3 += cv::Scalar(100,100,100);
    cv::Scalar b1 = cv::Scalar(70);
    cv::Scalar b3 = cv::Scalar(70,70,70);
    m1 = c*(m1 - b1);
    m3 = c*(m3 - b3);
    int m1_val = m1.at<uchar>(0,0);
    int m3_val = m3.at<cv::Vec3b>(0,0)[0];
    std::cout << c << " " << m1_val << " "  << m3_val << std::endl;
  }

  return 0;
}
```
Output:
```
c m1 m3
1 30 30
2 60 60
3 90 45 // For multi-channel 3*100 gets capped at 255. After subtraction of 3*70, result is 45
4 120 0 // For multi-channel 4*100 gets capped at 255. After subtraction of 4*70, result is 0
5 150 0
6 180 0
7 210 0
8 240 0
9 255 0 // The single-channel result is only capped once c*(30) is greater than 255
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [ ] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by dan-masek** [comment]
Yeah, there was definitely a lot of astonishment happening here, even with the understanding that some lazy evaluation was happening in the background.

The docs on the behaviour of matrix expressions are still rather sparse, and it seems at least in some cases the behaviour is not what you'd expect by simply reading the expression. Initially some warning may be in order.

**Comment by asmorkalov** [comment]
Related: https://github.com/opencv/opencv/issues/24908
